src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs

```diff
@@ -57,64 +48,111 @@ public string GenerateSchema()
+            AddDefaultConsoleLoggerOptionsSubSchema(context);
+        }
 
-            consoleLoggerOptionsSchema.Properties.Add(nameof(ConsoleLoggerOptions.FormatterName), formatterNameProperty);
-            consoleLoggerOptionsSchema.Properties.Add(nameof(ConsoleLoggerOptions.FormatterOptions), formatterOptionsProperty);
+        private static void AddConsoleLoggerOptionsSubSchema<TOptions>(GenerationContext context, ConsoleLoggerFormat consoleLoggerFormat)
+        {
+            JsonSchema consoleLoggerOptionsSchema = new JsonSchema();
             consoleLoggerOptionsSchema.RequiredProperties.Add(nameof(ConsoleLoggerOptions.FormatterName));
 
-            return consoleLoggerOptionsSchema;
+            JsonSchemaProperty formatterOptionsProperty = AddDiscriminatedSubSchema(
+                context.Schema.Definitions[nameof(ConsoleLoggerOptions)],
+                nameof(ConsoleLoggerOptions.FormatterName),
+                consoleLoggerFormat.ToString(),
+                nameof(ConsoleLoggerOptions.FormatterOptions),
+                consoleLoggerOptionsSchema);
+
+            formatterOptionsProperty.Reference = context.AddTypeIfNotExist<TOptions>();
         }
 
-        public static JsonSchema GenerateDefaultConsoleLoggerOptionsSchema(JsonSchema consoleFormatterOptions)
+        private static void AddDefaultConsoleLoggerOptionsSubSchema(GenerationContext context)
         {
             JsonSchema consoleLoggerOptionsSchema = new JsonSchema();
 
             JsonSchemaProperty formatterNameProperty = new JsonSchemaProperty();
             JsonSchemaProperty formatterOptionsProperty = new JsonSchemaProperty();
-            
-            JsonSchema formatterOptionsSchema = new JsonSchema();
-            formatterOptionsSchema.Reference = consoleFormatterOptions;
 
-            formatterOptionsProperty.Reference = formatterOptionsSchema;
+            formatterOptionsProperty.Reference = context.AddTypeIfNotExist<SimpleConsoleFormatterOptions>();
 
             formatterNameProperty.Type = JsonObjectType.Null;
             formatterNameProperty.Default = "Simple";
 
             consoleLoggerOptionsSchema.Properties.Add(nameof(ConsoleLoggerOptions.FormatterName), formatterNameProperty);
             consoleLoggerOptionsSchema.Properties.Add(nameof(ConsoleLoggerOptions.FormatterOptions), formatterOptionsProperty);
 
-            return consoleLoggerOptionsSchema;
+            context.Schema.Definitions[nameof(ConsoleLoggerOptions)].OneOf.Add(consoleLoggerOptionsSchema);
+        }
+
+        private static JsonSchemaProperty AddDiscriminatedSubSchema(
+            JsonSchema parentSchema,
+            string discriminatingPropertyName,
+            string discriminatingPropertyValue,
+            string discriminatedPropertyName,
+            JsonSchema subSchema = null)
+        {
+            if (null == subSchema)
+            {
+                subSchema = new JsonSchema();
```

Comment:
Not sure if this is actually a problem, but I noticed that you separated `FormatterName` as a RequiredProperty (line 69) from where you add `discriminatedPropertyName` as a property down below. In theory, if `subSchema` is null, then when we initialize `subSchema` we won't have `FormatterName` as a required property. Do you think this is a problem? Should we add `subSchema.RequiredProperties.Add(nameof(ConsoleLoggerOptions.FormatterName))` on line 120?

Reply:
I don't think that's going to be an issue, in fact, the "required" array on the sub schemas in this case might not actually be needed at all (for any of the given subschemas, if FormatterName is "const": "<somevalue>", then by virtue of the fact that it has a value then the requiredness of the property is fulfilled implicitly).

There will be other schemas where the properties in the parent schema are already defined to be required. It _might_ not be necessary to redeclare that those properties in the subschema are again required. My empirical testing has not shown that this is necessary.